### ISSUE TITLE
can't retrieve non-text files using `preston get`

### ISSUE THREAD

**Issue by mielliott** [bio-guoda/preston#100]
I haven't tested it, but I believe the recent changes to `preston get` (allowing it to accept content-based locations instead of just content hashes as input) mean that it can only retrieve text data, and will fail to retrieve things like archives, compressed files, media files, etc.. I'll verify and fix this when I get a chance. fyi @jhpoelen

**Comment by jhpoelen** [comment]
I can confirm that:

```
$ java -jar preston.jar get 'hash://sha256/94a4fc4824d951c55860e3e5c4c662afcfae55d52a25105bca1cb6a7b3a062' | sha256sum
94a4fc4824d951c0155860e3e5c4c662afcfae55d52a25105bca1cb6a7b3a062  -
```
with preston v0.0.10 (the one that comes with https://github.com/bio-guoda/preston-amazon)

however, when running v0.2.3 :
```
 preston get 'hash://sha256/94a4fc4824d951c0155860e3e5c4c662afcfae55d52a25105bca1cb6a7b3a062' | sha256sum
java.io.IOException: problem retrieving [hash://sha256/94a4fc4824d951c0155860e3e5c4c662afcfae55d52a25105bca1cb6a7b3a062]
	at bio.guoda.preston.cmd.CmdGet.handleContentIri(CmdGet.java:73)
	at bio.guoda.preston.cmd.CmdGet.run(CmdGet.java:51)
	at bio.guoda.preston.cmd.CmdGet.run(CmdGet.java:38)
	at bio.guoda.preston.cmd.CmdLine.run(CmdLine.java:18)
	at bio.guoda.preston.cmd.CmdLine.run(CmdLine.java:26)
	at bio.guoda.preston.Preston.main(Preston.java:19)
Caused by: java.io.IOException: failed to resolve to content
	at bio.guoda.preston.cmd.CmdGet$ContentQuery.dereference(CmdGet.java:111)
	at bio.guoda.preston.cmd.CmdGet.processQuery(CmdGet.java:78)
	at bio.guoda.preston.cmd.CmdGet.handleContentIri(CmdGet.java:70)
	... 5 more
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  -
```
with 

```
$ preston version
0.2.3
```

**Comment by jhpoelen** [comment]
However, when running:

```
$ preston get 'zip:hash://sha256/94a4fc4824d951c0155860e3e5c4c662afcfae55d52a25105bca1cb6a7b3a062!/eml.xml' 
<?xml version='1.0' encoding='utf-8'?><eml:eml xmlns:eml="eml://ecoinformatics.org/eml-2.0.1" xmlns:md="eml://ecoinformatics.org/methods-2.0.1" xmlns:proj="eml://ecoinformatics.org/project-2.0.1" xmlns:d="eml://ecoinformatics.org/dataset-2.0.1" xmlns:res="eml://ecoinformatics.org/resource-2.0.1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:dc="http://purl.org/dc/terms/" system="Plazi" scope="system" packageId="FFF76241FFB3B773A711FF98DF6BFFF3/eml-1519775222262.xml"><dataset><alternateIdentifier>FFF76241FFB3B773A711FF98DF6BFFF3</alternateIdentifier><alternateIdentifier>http://dx.doi.org/10.3897/zookeys.740.14640</alternateIdentifier><a
```

content is returned as expected. . .

**Comment by mielliott** [comment]
It's fixed now:
```
$ preston get 'hash://sha256/94a4fc4824d951c0155860e3e5c4c662afcfae55d52a25105bca1cb6a7b3a062' | sha256sum
94a4fc4824d951c0155860e3e5c4c662afcfae55d52a25105bca1cb6a7b3a062  -
```

**Comment by jhpoelen** [comment]
@mielliott Thanks for fixing and for adding the test cases!

The ContentQuery implementation of the CmdGet class looks pretty neat.

In a way it is a content resolver (dereferencer) that uses a content-based universe (content hash, blobstore) instead of the location-based universe (url and http client).

Did you consider to make the ContentQuery implement the Dereferencer<InputStream> interface?

**Comment by mielliott** [comment]
> Did you consider to make the ContentQuery implement the Dereferencer<InputStream> interface?

Ooh, good idea. The suggested change is implemented in bf3465cca131c5781dcce3cec2aaa880859a7941.

I've also put the "get bytes identified by a content query" logic into it's own ContentDerefencer class (see bf4d2aa6fa70b4c066079e387949cf6a483659d4). I feel like this could be useful for all sorts of applications.
